elements/dash-tabs/dash-tabs.js

```diff
@@ -0,0 +1,293 @@
+   * All children of `<dash-tabs>` should be either `<dash-tab>` or
+   * `<dash-tabpanel>`. This element is stateless, meaning that no values are
+   * cached and therefore, changes during runtime work.
+   *
+   * Because both the tabs and the panels link to each other semantically, every
+   * tab and every panel _must_ have a unique ID.
+   */
+  class DashTabs extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    /**
+     * `connectedCallback` groups tabs and panels by reordering and makes sure
+     * exactly one tab is active.
+     */
+    connectedCallback() {
+      // The element needs to do some manual input event handling to allow
+      // switching with arrow keys and Home/End.
+      this.addEventListener('keydown', this._onKeyDown);
+      this.addEventListener('click', this._onClick);
+
+      this.setAttribute('role', 'tablist');
+      // Acquire all tabs and panels inside the element
+      const tabs = this._allTabs();
+      const panels = this._allPanels();
+      // If there are no tabs, there is no way to switch between panels. Abort.
+      if (!tabs.length) return;
+
+      // For progressive enhancement, the markup should alternate between tabs
+      // and panels. If JavaScript is disabled, all panels are
+      // visible with their respective tab right above them.
+      // If JavaScript is enabled, the element groups all children by type.
+      // First all the tabs, then all the panels.
+      // Calling `appendChild` on an already inserted element _moves_ the
+      // element to the last position.
+      tabs.forEach(tab => this.appendChild(tab));
+      panels.forEach(panel => this.appendChild(panel));
+
+      // Give each panel a `aria-labelledby` attribute that refers to the tab
+      // that controls it.
+      tabs.forEach(tab => {
+        if (tab.id)
+          this._panelForTab(tab).setAttribute('aria-labelledby', tab.id);
+        else
+          console.warn(`The tab that controls panel #${panelId} has no id, ` +
+            `but needs one for ARIA.`);
+      });
+
+      // The element checks if any of the tabs have been marked as selected. If
+      // not, the first tab is now selected.
+      const selectedTab =
+        tabs.find(tab =>
+          tab.getAttribute('aria-selected') === 'true') || tabs[0];
+
+      // Next, we switch to the selected tab. `selectTab` takes care of marking
+      // all other tabs as deselected and hiding all other panels.
+      this._selectTab(selectedTab);
+    }
+
+    /**
+     * `_allPanels` returns all the panels in the tab panel. This function could
+     * memoize the result if the DOM queries ever become a performance issue.
+     * The downside of memoization is that dynamically added tabs and panels
+     * will not be handled.
+     *
+     * This is a method and not a getter, because a getter implies that it is
+     * cheap to read.
+     */
+    _allPanels() {
+      return Array.from(this.querySelectorAll('dash-tabpanel'));
```

Comment:
Yes, `NodeList` has `forEach` in Chrome, Safari and Firefox, but not `find`, `findIndex` or `filter`. So I’d rather convert to a proper array.